models/event.js

```diff
@@ -104,21 +82,75 @@ function buildWhereClause(where, nextArgumentIndex) {
   const values = [];
   const conditions = Object.entries(where).map(([key, value]) => {
     values.push(value);
-    const column = columnMap[key] ?? key;
+    let column;
+
+    if (key.startsWith('metadata.')) {
+      const metadataField = key.substring(key.indexOf('.') + 1);
+      column = `events.metadata->>'${metadataField}'`;
+    } else {
+      column = `events.${columnMap[key] ?? key}`;
+    }
+
     return Array.isArray(value) ? `${column} = ANY ($${nextArgumentIndex++})` : `${column} = $${nextArgumentIndex++}`;
   });
 
   return {
-    text: `${conditions.join(' AND ')}`,
+    text: conditions.length ? `WHERE ${conditions.join(' AND ')}` : '',
     values: values,
     nextArgumentIndex: nextArgumentIndex,
   };
 }
 
+async function findAllRelatedEvents(id) {
+  const query = {
+    text: `
+    WITH searched_event AS (
+      SELECT
+        id,
+        metadata
+      FROM
+        events
+      WHERE
+        id = $1
+      LIMIT
+        1
+    ),
+    same_metadata_events AS (
+      SELECT
+        ARRAY_AGG(events.id) AS ids
+      FROM
+        events
+      INNER JOIN
+        searched_event ON true
+      WHERE
+        events.metadata @> searched_event.metadata
```

Comment:
Do jeito que está testando, também passaria se fizesse a comparação simplesmente com `=`:

```sql
WHERE
  events.metadata = searched_event.metadata
```

Mas acredito que a intenção seja lidar com os casos em que nem todos os afetados serão os mesmos, por exemplo quando um conteúdo (ou usuário) for pego no primeiro evento, mas, por já ser mais antigo, não for pego pelo segundo. Nesse caso, para obter o mesmo resultado independentemente de qual dos eventos enviar na requisição, a comparação precisa ser com algo assim:

```sql
WHERE
  events.metadata @> searched_event.metadata OR
  events.metadata <@ searched_event.metadata
```

Obs.: Tem outras formas de comparar, mas teria que ter algo específico para conteúdos e outra comparação para usuários.

Acho que, para testar isso corretamente, teria que ser possível criar os eventos com data retroativa. Uma alternativa é modificar a query de criação do evento para usar a data e hora da lambda ao invés do banco de dados (como no `published_at`) e usar `vi.useFakeTimers` no teste.

Reply:
Fiquei em dúvida se vale a pena tantas alterações para testar isso. Veja o que fui percebendo:

1. Ao passar um `new Date()` no `INSERT` do `models/event`, não será possível testar com o `RequestBuilder` porque o Next e o Vitest são processos diferentes, então mesmo modificando o tempo no teste, o registro usará o horário do computador.
2. Então eu poderia usar o `orchestrator` para criar os conteúdos, já que ele cria também o evento. Nessa situação, o horário salvo é o modificado no teste com `useFakeTimers()`, mas o `canRequest` não é chamado.
3. Seguindo, eu precisaria modificar o `orchestrator.createContent` para simular a chamada à regra adequada do firewall.
4. Por fim (imagino), precisaria modificar a função `firewall_create_content_text_child` e relacionadas para aceitar o `createdAt`, que seria usado para a comparação no lugar de `AND created_at > NOW() - INTERVAL '5 seconds'`. 

Faz sentido? Compliquei alguma parte que não precisaria?

Reply:
> 1. Ao passar um `new Date()` no `INSERT` do `models/event`, não será possível testar com o `RequestBuilder` porque o Next e o Vitest são processos diferentes, então mesmo modificando o tempo no teste, o registro usará o horário do computador.
> 2. Então eu poderia usar o `orchestrator` para criar os conteúdos, já que ele cria também o evento. Nessa situação, o horário salvo é o modificado no teste com `useFakeTimers()`, mas o `canRequest` não é chamado.

Para usar tempos falsos, de fato precisa criar o conteúdo pelo `orchestrator.createContent`, mas só precisa falsificar a hora do primeiro (ou dos primeiros) evento(s). A requisição que vai cair no firewall deve ser direto pela API, onde pode ser usado o `RequestBuilder`.

> 3. Seguindo, eu precisaria modificar o `orchestrator.createContent` para simular a chamada à regra adequada do firewall.

Acho que não precisa (nem deveria fazer isso), mas entendo que isso seria para a criação do primeiro evento de firewall do teste. Deve ser suficiente adicionar um `orchestrator.createFirewallEvent` para permitir criar um evento qualquer diretamente por ele. Com isso será possível simular um evento de firewall com mais (e com menos) conteúdos (ou usuários) do que o evento de firewall que será criado pela API, e daí é esperado que esse evento criado manualmente seja incluído nos eventos de firewall relacionados que são retornados. Acho que compensa fazer isso, pois a alternativa seria criar os dois eventos pela API, mas adicionando um intervalo de tempo entre as duas chamadas, o que aos poucos vai atrasando a execução da bateria de testes

Sobre o `orchestrator.createFirewallEvent`, ele pode chamar um `firewall.createEvent`, que pode ser criado para ser usado dentro de `firewall/rules`, o que também parece ser uma refatoração simples.

> 4. Por fim (imagino), precisaria modificar a função `firewall_create_content_text_child` e relacionadas para aceitar o `createdAt`, que seria usado para a comparação no lugar de `AND created_at > NOW() - INTERVAL '5 seconds'`.

Pensando rapidamente, acho que não precisa modificar essas funções.

> Faz sentido? Compliquei alguma parte que não precisaria?

Talvez tenha complicado ao pensar em usar apenas o `orchestrator.createContent` ou apenas o `RequestBuilder`, mas usando os dois, cada um no momento adequado, acho que é uma mudança bem simples. No `Triple A` dos testes (`Arrange`, `Act`, `Assert`), o `orchestrator.createContent` e `orchestrator.createFirewallEvent` seriam usados no `Arrange` e o `RequestBuilder` no `Act`.

No código atual em produção seria uma mudança apenas na `query` de `event.create`, por exemplo adicionando `created_at`, `$5` e `new Date()` assim:

```js
  const query = {
    text: `INSERT INTO events (type, originator_user_id, originator_ip, metadata, created_at)
               VALUES($1, $2, $3, $4, $5) RETURNING *;`,
    values: [
      cleanObject.type,
      cleanObject.originator_user_id,
      cleanObject.originator_ip,
      cleanObject.metadata,
      new Date(),
    ],
  };
```

Não testei o que estou sugerindo, então posso estar deixando escapar algo. Se preferir, posso implementar para descobrir se é tão simples como pensei. 🤝

Reply:
@aprendendofelipe Eu estava tentando implementar mas comecei a ficar confuso sobre o que você realmente propôs, então se puder implementar, agradeço.

A alteração que subi agora foi apenas para definir o `status: published` de alguns testes que criavam publicações pelo `orchestrator`.

Reply:
Pode deixar que amanhã eu implemento 🤝

Reply:
As alterações que subi agora foram apenas um `rebase`.

Reply:
Criei os testes alterando a data de criação do evento, foi a forma mais simples que encontrei. Além disso, fiz alguns ajustes conforme necessário.

Acha que o comportamento precisa ser modificado ou faz sentido como está?